### ISSUE TITLE
cURL add support for SSL options

### ISSUE THREAD

**Issue by adpe** [maths/moodle-qtype_stack#412]
Hello @sangwinc 

For run the MaximaPool over secure mode (SSL) you must give some parameters, that cURL can successfully verify the certificate.

Something similar I found in the `.../moodle/search/engine/solr` setup. Here's the configuration for the connection:
![image](https://user-images.githubusercontent.com/9467708/47548558-c9dd9b00-d8f9-11e8-9192-d49573209c1e.png)

At the moment for my case it's enough to just add a new settings field like:
![image](https://user-images.githubusercontent.com/9467708/47548668-f5f91c00-d8f9-11e8-887b-4c1db5a4e37e.png)

But you must be aware, that when you uncheck this field, it can be a risk. But what's more risky? HTTP or HTTPS without certificate validation? If we want to check the certificate we must add more cURL options to get it working.

What do you thought about this case?
Adrian

**Comment by aharjula** [comment]
Well, one might also think about why one needs HTTPS for MaximaPool at all? Essentially, the Pool should never be directly exposed to the public Internet and the only connections it should receive should be through internal networks and even in those cases only from the Moodle/ILIAS servers running STACK (+ maybe some admin users). Within such network setup it should not matter whether the information in transit is readable or even modifiable as you are working in a trusted context.

If one needs/wants to run SSL then one can simply modify the STACK side connector-class to set those settings for cURL. But personally, I would consider SSL as wasted resources for MaximaPools in internal networks and firewall controlled access policies. But if you add some form of authentication to your Pool and place it into a network where unknown people can see and access it then you might need that, however if you go for such a setup you are essentially asking for someone to crack it.

Also SSL handshakes take time and MaximaPool exists to save time, so if possible keep unnecessary time use to a minimum and skip SSL if at all possible.

**Comment by adpe** [comment]
Hi @aharjula 

Yes that's right. In most cases it makes not much sense, but there can be individual cases. I've seen, that with SSL the performance is lower.

I'll look how to deal with it and our `Security officer`.

Nevertheless, I will publish my solution.

**Comment by timhunt** [comment]
I don't understand why the Solr config screen requires so many millions of settings. HTTPS URLS should just work in most places.

If you are setting up https, why not use a proper certificate?

If we consider it essential to make self-signed certificates work, the a single option should be enough. However, I think the config variable should have its meaning inverted, so it fails safe. E.g. call it casskipmaximaservercertverfiy.

get_cas_verifypeer_value function seems to be taking half a dozen lines of code to do something trivial. With the above change of setting name, it just becomes return empty(self::$config->casskipmaximaservercertverfiy)

Normally, if you are turning off CURLOPT_SSL_VERIFYPEER you also turn off CURLOPT_SSL_VERIFYHOST, at least in other code I have seen.

**Comment by aharjula** [comment]
Like Tim I wonder why not use a proper certificate if you must use SSL. Also I strongly believe that SSL is not a thing that belongs in the same picture with MaximaPool as it focuses on securing stuff that matters not when the MaximaPool is correctly secured by cordoning it out of the world. Also setting that option for cURL essentially removes quite a lot of the security that SSL would bring to the picture, essentially I consider SSL when mixed with MaximaPool to be part of a bad form of _security theater_ and toggling that setting would make it very pointless theater indeed.

Actually, having though about this for a while now I think that having an option to set that cURL setting might send a bad message. Either by making people think that SSL has any role here or that setting that kind of an security feature off is a valid thing to do.

**Comment by adpe** [comment]
Thanks for your replies.

Internal we have our own CA where we sign our certificates, so we have no proper trusted certificate authority - so, internal you can handle that by deposit the root certificate. For this case, you should then provide the key files - like in Solr.

So as I understood, there should be two things:
1. If the user is using HTTPS in the `Maxima command` it should show the user a message info, that's not working in due of a certificate error (e.g self-signed).
2. As solution on own-risk, the user can turn off `CURLOPT_SSL_VERIFYPEER` and `CURLOPT_SSL_VERIFYHOST` with the `casskipmaximaservercertverfiy` checkbox to get it work.

Please correct me, if I'm wrong.

**Comment by aharjula** [comment]
My view is that your point 1. is a valid point and should indeed happen. But I do not like the second point as it brings up flashbacks from my studies...

If you have your own CA you might as well add that CA:s certs to the CA-store installation on the server so that cURL can find them and actually make SSL do what it is supposed to do. Based on what I learned on my computer security classes SSL/TLS only secures the transport of data and is intended to ensure that there is no man in the middle modifying it but if you drop that verification it cannot do that unless you go for a PSK based solution and that would again mean more config options. Personally, I am not even happy about the extra options we already have for HTTP "authentication" as it too basically does nothing (without SSL that is, and I do not believe in SSL in this use case) so maybe I am being difficult.

Essentially, if I have understood correctly SSL without verification is merely a way of obfuscation and obfuscation does not bring security and I would prefer not to provide means to skip that verification just for conveniences sake as it skips the security at the same time. If someone wants or needs to use SSL they might as well do it correctly, although I fear that most that think that sprinkling SSL over things will bring security, will also simply find that verification disabling option from stackexchange or some such and select it without understanding what it means.

Yet again SSL protects you from MITM attacks, and if there is a risk for such in your MaximaPool <-> Moodle server connections then you might want consider what you are doing and could you do something differently to avoid SSL and/or MITM attacks. But if those attacks are a possibility you should never disable that verification. Also positively identifying suitable clients (Moodle servers) and servers (MaximaPool servers) is of great importance if you have a situation where MITM would be a risk, but for that you would probably go for a PSK solution.

**Comment by adpe** [comment]
Yes you're right @aharjula. 

When there is nothing specified, cURL takes the default `cainfo` of the OS. I just checked this in our CentOS and it should work with our own CA certificate, you must just update your `ca-trust`. So, I must correct my sentence from above `For this case, you should then provide the key files - like in Solr.` **Isn't correct!**.

Finally, the problem is "solved". Just using HTTPS with a non valid certificate leads to an error. Here it would be great, if it would show an info message. But it's not really necessary.

Let me know what's to do. If no action is needed, you can close this issue.

**Comment by sangwinc** [comment]
Adrian, thank you very much indeed for contributing here. 
I really don't know enough about SSL, HTTPS and the rest of the secure server management to make an informed contribution to this issue!  So, should this issue remain open, and what should we do about the pull request?

**Comment by aharjula** [comment]
My opinion is that we should not provide means to disable security to people seeking security from SSL (which just by itself provides no real security in this case) and I would drop both the issue and pull request. However, if the pull request contains a warning message logic for expressing that cURL cannot validate the certificates and fails then we should merge that bit of it as it is a common issue. Just to repeat myself while it is a common issue, turning validation off is not the solution and all those people giving that as a "solution" on various sites are either evil, don't know what they are talking about or make wild assumptions about how SSL is deployed in a given case and leave those assumptions undocumented.

I don't know enough either but I did study securing information systems under a professor that actually  took part in designing large scale real world crypto systems like GSM and I do remember some stuff from there.

So Chris, Adrian's issue was solved by letting cURL have the information it needed through the correct channels i.e. the `cainfo` and things are therefore ok. Only open issue here is the possible error message to give a hint to the user that cURL can't do what it needs and the issue is this.
